PM-26250 Explore options to enable direct importer for mac app store build

harr1424 opened the pull request:
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-26250?atlOrigin=eyJpIjoiN2Y1Y2NhNmQ4OTBhNDU4YzkwZWNmYmI1MTdiNzc1NDgiLCJwIjoiaiJ9

## 📔 Objective

For the direct importer, browser profiles aren't populating when the desktop app was installed via the mac app store. This is because macOS apps can be [sandboxed](https://developer.apple.com/documentation/xcode/configuring-the-macos-app-sandbox) and won't have access to system directories to look for browser profiles.

The goal of this effort is to enable the direct importer for the mac app store build.

## 🎥  Screencast

![Screen Recording 2025-11-21 at 14 32 37 mov](https://github.com/user-attachments/assets/1c0a716d-6273-4f50-921a-0f459f6075e0)


## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

harr1424 committed:
initial sandbox implementation for mac app store builds

harr1424 committed:
chromium importer working in sandbox

harr1424 committed:
add import-chrome-component.ts to working tree

harr1424 committed:
move sandbox code into macos.rs

harr1424 committed:
add/improve comments and logging

harr1424 committed:
added docs and removed debug output

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
remove additional debug output and revert popup-modal-style to match main

harr1424 committed:
Merge branch 'PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build' of github.com:bitwarden/clients into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

merge conflict resolution

harr1424 committed:
add dep to desktop_native cargo.toml

harr1424 committed:
cargo fmt

harr1424 committed:
avoid unwrap

harr1424 committed:
pass args directly in objc/build.rs

harr1424 committed:
cargo clippy

harr1424 committed:
fix another clippy warning

harr1424 committed:
remove unused dep

harr1424 committed:
fix more clippy lints not caught locally

harr1424 committed:
re-run cargo fmt after fixing clippy lints

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
replace swift logic with objc

harr1424 committed:
Merge branch 'PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build' of github.com:bitwarden/clients into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

merge main

harr1424 committed:
rename files to avoid uppercase letters

harr1424 committed:
rename references

coroiu commented on apps/desktop/desktop_native/objc/build.rs:
@coltonhurst You're currently the only users of the `objc` crate and its `runCommand` framework. In hindsight I admit that the JSON approach might not have been the best, but there's still some callback-to-async and string interop for automatically freeing memory (without depending on `libc`) that I feel is worth using/saving. Does Autofill have any current plans for improving the `objc` crate? Would you also prefer swift?

coltonhurst replied:
@coroiu thanks for the mention. I am not intimately familiar with this area of the code base, but I agree with your points.

I think keeping only Objective-C code in the `objc` directory is a good move. It looks like that might have been updated in this PR since your message and my time in looking at this? (Thanks for your patience!)

Regarding the JSON pattern... I don't think we should deviate from an existing pattern, creating two different patterns, unless there is a plan to transition the old stuff, or if there is a specific reason we need to do it a different way. This is just my preference; I would rather code patterns be consistent (even if not the "perfect" solution), rather than "right" stuff that is not consistent and less easily understood to new readers. So... the tldr for this point is it might be worth looking into a way to do this in a way that follows the existing pattern, rather than creating new patterns.

coltonhurst replied:
Additionally, we have far reaching plans for significant desktop_native improvements as a whole. However, nothing will happen in an immediate time frame.

coltonhurst replied:
@harr1424 do you mind letting me know when the files relevant to this convo are ready for review? (That may be now, I'm not sure.)

I also think it would be nice to also add some architecture documentation to our contributing docs around the pattern so it's at least documented more. Any ideas @coroiu?

coltonhurst replied:
@harr1424 no problem! I don't think I'm a required code owner anyways, just was getting involved!

harr1424 committed:
use progress spinner for imports

harr1424 committed:
improved error display

Hinton commented on apps/desktop/desktop_native/objc/src/native/chromium_importer/browser_access_manager.m:
```diff
@@ -0,0 +1,196 @@
+        };
+    }
+    return self;
+}
```
I think if we want we can keep this information on the rust side and minimize the obj-c we have to maintain.

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
respond to review comments:

• replace usage of compile time flag with boolean for conditinal logic
• moved sandbox specific logic to contained module inside macos.rs
• remove redundant browser array from objc code and pass the target browser path as arg

harr1424 committed:
pass new bool value on other platforms

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
remove sandbox feature flag and refactor to use bool

harr1424 committed:
fix cargo fmt

harr1424 committed:
fix clippy error

harr1424 committed:
fix unused variable outside of macos build

harr1424 committed:
fix lints on linux build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
eval Claude's comments and address legitimate findings

harr1424 committed:
clippy lints

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
remove callout and revert to previous error display

harr1424 committed:
Merge branch 'PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build' of github.com:bitwarden/clients into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

pull latest from main

harr1424 committed:
adapt chromium_importer to match conventions established  by autofill:
• eliminate extern c and unsafe rust from chromium_importer logic (still called in prior napi/lib.rs)
• use CommandInput / CommandResult JSON pattern to call objc methods
• remove usage of libc to cleanup memory and remove this dependency

harr1424 committed:
clippy & fmt

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
remove ability for user to select dir different to browser

harr1424 committed:
Merge branch 'PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build' of github.com:bitwarden/clients into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

merge latest

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
use new function name introduced in merge

harr1424 committed:
disable clippy warning on conditionally compiled code block

harr1424 committed:
fix breaking changes caused by PM-27081

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
fix more lints

harr1424 committed:
rework error handling & presentation

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
use bit-spinner instead of bwi-spin

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
remove comment related to sandbox flag (outdated)

Hinton commented on apps/desktop/desktop_native/chromium_importer/src/chromium/platform/macos.rs:
```diff
@@ -7,6 +7,220 @@ use crate::{
+        let bundle_id = match bundle_id {
+            Some(id) => id,
+            None => return Ok(true),
+        };
```
**issue:** Use `ok_or`, https://contributing.bitwarden.com/contributing/code-style/rust#ok_or.

Hinton commented on apps/desktop/desktop_native/build.js:
**issue:** Revert.

Hinton reviewed (commented):
Only reviewed the native code so far.

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
address review comments

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

coroiu commented on apps/desktop/desktop_native/chromium_importer/src/chromium/platform/macos.rs:
```diff
@@ -7,6 +7,220 @@ use crate::{
+                }),
+            };
+
+            let output = desktop_objc::run_command(serde_json::to_string(&input)?)
```
The original use-case makes the types opaque to Rust and are instead defined in TypeScript. Ideally I would've like for autofill to take the lead on thise since they are the ones that primarily use this and know how such a change will affect their code. But I agree that the way we're using it here, constructing the json in rust, it would be better with defining structs

coroiu reviewed (dismissed):
Looks good (no big changes),please hold off on pinging me from further reviews until you have approval from @Hinton so I don't have to approve an empty change set :)

Hinton commented on apps/desktop/desktop_native/chromium_importer/src/chromium/platform/sandbox.rs:
```diff
@@ -0,0 +1,206 @@
+    Success { value: T },
+    #[serde(rename_all = "camelCase")]
+    Error { error: String },
+}
```
```suggestion
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", tag = "type")]
enum CommandResult<T> {
    Success { value: T },
    Error { error: String },
}
```

Hinton commented on apps/desktop/desktop_native/chromium_importer/src/chromium/platform/sandbox.rs:
```diff
@@ -0,0 +1,206 @@
+struct CheckBrowserInstalledResponse {
+    #[serde(rename = "isInstalled")]
+    is_installed: bool,
+}
```
```suggestion
#[derive(Debug, Deserialize)] 
#[serde(rename_all = "camelCase")]
struct CheckBrowserInstalledResponse {
    is_installed: bool,
}
```

Hinton commented on apps/desktop/src/app/tools/import/import-desktop.component.ts:
```diff
@@ -39,7 +47,13 @@ export class ImportDesktopComponent {
 
-  constructor(public dialogRef: DialogRef) {}
+  protected readonly onLoadProfilesFromBrowser = this._onLoadProfilesFromBrowser.bind(this);
+  protected readonly onImportFromBrowser = this._onImportFromBrowser.bind(this);
```
**question:** What's up with the bind here?

Hinton replied:
Passing callbacks as inputs is generally a sign of something being wrong. I think in this case we would probably be better of making a service that abstracts this than injecting callbacks. Probably outside of the scope of this ticket though.

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
revert breaking changes from review suggestions

harr1424 committed:
fix syntax error

harr1424 committed:
cargo fmt

harr1424 committed:
safe changes related to review feedback (checkpoint)

harr1424 commented on apps/desktop/desktop_native/objc/src/native/chromium_importer/browser_access_manager.m:
```diff
@@ -0,0 +1,157 @@
+    if (![[NSFileManager defaultManager] fileExistsAtPath:localStatePath.path]) {
+        // Invalid directory selected caller will handle
+        return nil;
+    }
```
Sorry, any directory containing a `Localstate` file will technically pass, I will make this more defensive

harr1424 commented on apps/desktop/desktop_native/objc/src/native/chromium_importer/browser_access_manager.m:
```diff
@@ -0,0 +1,157 @@
+                                            relativeToURL:nil
+                                            error:&error];
+
+    if (!bookmarkData) {
```
Why would assigning `nil` to this variable be helpful here?

harr1424 committed:
address remaining review comments

Hinton commented on libs/importer/src/components/chrome/import-chrome.component.ts:
```diff
@@ -105,15 +105,29 @@ export class ImportChromeComponent implements OnInit, OnDestroy {
     effect(async () => {
-      this.profileList = await this.onLoadProfilesFromBrowser(this.getBrowserName(this.format()));
-      // FIXME: Add error handling and display when profiles could not be loaded/retrieved
+      // Callback is set via @Input after constructor, so check it exists
```
**issue:** Effect is used when you want to react to a signal being changed. There is no signal here.

> Registers an "effect" that will be scheduled & executed whenever the signals that it reads changes.

You can use `ngOnInit`, or convert the inputs to a signal.

harr1424 commented on apps/desktop/src/app/tools/import/import-desktop.component.html:
```diff
@@ -12,7 +12,7 @@
       @if (loading) {
         <div class="tw-absolute tw-inset-0 tw-flex tw-items-center tw-justify-center">
-          <i class="bwi bwi-spinner bwi-spin bwi-3x tw-text-primary-600" aria-hidden="true"></i>
+          <bit-spinner></bit-spinner>
```
@sukhleenb do you have a preference as to which approach is used here? 

1. Limiting the spinner to the submit button
2. Displaying the spinner in the dialog box as shown [here](https://private-user-images.githubusercontent.com/84741727/519303473-5ce79dd8-4a89-4a2f-928e-cdc753e5cc96.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjU4OTk2MjUsIm5iZiI6MTc2NTg5OTMyNSwicGF0aCI6Ii84NDc0MTcyNy81MTkzMDM0NzMtNWNlNzlkZDgtNGE4OS00YTJmLTkyOGUtY2RjNzUzZTVjYzk2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE2VDE1MzUyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjOGY3YWIzNTY5Njc5MmIwNmM3NGMwZTEzNDFmNDdiMTgwMmY0NDI5YTg1MmM5MjUxYmRlMWRmZmQ0OGNiNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cGfy1qNlPFgE2jp2LdExyc8XtcsTeQLa4nKY7IJKK4Q)

harr1424 commented on apps/desktop/desktop_native/chromium_importer/src/chromium/platform/sandbox.rs:
```diff
@@ -0,0 +1,207 @@
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+enum CommandResult<T> {
+    // rename = "camelCase" was a review suggestion with breaking changes
```
Sorry, this outdated and incorrect comment has been removed. It was actually your suggested use of `ok_or()` that had breaking changes: 

```rust
async fn is_browser_installed(browser_name: &str) -> Result<bool> {
    let bundle_id = BROWSER_BUNDLE_IDS
        .iter()
        .find(|(name, _)| *name == browser_name)
        .map(|(_, id)| *id);

    let Some(bundle_id) = bundle_id else {
        return Ok(true); // ok_or(true) was a review suggestion with breaking changes
                         // Avoid ok_or(true): serializes Result as {"Ok": "..."} instead of string
                         // value
    };
    ```

harr1424 commented on apps/desktop/src/app/tools/import/import-desktop.component.ts:
```diff
@@ -48,11 +62,37 @@ export class ImportDesktopComponent {
+      // Check verbose error chain for specific i18n key indicating browser not installed
+      const browserNotInstalledMatch = rawMessage.match(
+        /chromiumImporterBrowserNotInstalled:([^:]+)/,
+      );
```
Not an enum exactly but `i18n` translation keys: 

```json
  "chromiumImporterBrowserNotInstalled": {
    "message": "$BROWSER$ is not installed. Please install $BROWSER$ before importing passwords from it.",
    "placeholders": {
      "browser": {
        "content": "$1",
        "example": "Chrome"
      }
    }
  },  
    "browserAccessDenied": {
    "message": "Either an invalid browser directory was selected, or access was denied"
  },
  ```

Hinton replied:
Enum from the rust side which is then mapped to the appropriate i18n key in the client.

harr1424 committed:
address review comments

Hinton commented on apps/desktop/src/app/tools/import/desktop-import-metadata.service.ts:
```diff
@@ -20,7 +20,8 @@ export class DesktopImportMetadataService
 
   async init(): Promise<void> {
-    const metadata = await ipc.tools.chromiumImporter.getMetadata();
+    const isMas = ipc.platform.isMacAppStore;
```
Yes `platformUtilsService.isMacAppStore`. `ipc.platform.isMacAppStore` should probably be considered internal to the platform team.

itsadrago replied:
Is `this.system.environment.isMacAppStore()` the correct replacement then?

harr1424 replied:
```typescript 
export class DesktopImportMetadataService
  extends DefaultImportMetadataService
  implements ImportMetadataServiceAbstraction
{
  constructor(system: SystemServiceProvider) {  // note system refers to an instance of SystemServiceProvider
    super(system);
  }

...

    const isMas = this.system.environment.isMacAppStore();

```

```typescript
export type SystemServiceProvider = {
  /** Policy configured by the administrative console */
  readonly policy: PolicyService;

  /** Client extension metadata and profile access */
  readonly extension: ExtensionService;

  /** Event monitoring and diagnostic interfaces */
  readonly log: LogProvider;

  /** Config Service to determine flag features */
  readonly configService: ConfigService;

  /** Platform Service to inspect runtime environment */
  readonly environment: PlatformUtilsService;    // note that environment refers to an instance of PlatformUtilsService

  /** SDK Service */
  readonly sdk?: BitwardenClient;
};
```

Therefore `this.system.environment` should be considered synonymous with `PlatformUtilsService` in the context of this file.

Hinton commented on apps/desktop/src/app/tools/import/chromium-importer.service.ts:
**issue:** This file is located in the wrong directory. `app` contains renderer concerns but this uses `ipcMain` i.e. it's a main service.

itsadrago replied:
LGTM

Hinton reviewed (commented):
Outstanding comment besides the location of the main service below,  https://github.com/bitwarden/clients/pull/17479#discussion_r2611145565.

The other comments are future improvements and non blocking.

harr1424 committed:
address review comments

harr1424 committed:
resolve conflict with main branch

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

coroiu reviewed (dismissed):
I think `desktop_native/objc/src/native/chromium_importer` should be added to codeowners so it doesn't fall under platform ownership, but I'm fine with doing that in a a follow-up PR, other than that there aren't many platform changes so I'll approve

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
restore mas_build arg after napi split

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

coroiu reviewed (dismissed):
No real changes to platform code

harr1424 committed:
merge latest main

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
resolve merge conflicts

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
resolve conflicts and ai review

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
use sandboxed dir for path resolution

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
remove empty clean file

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
Merge branch 'main' into PM-26250-Explore-options-to-enable-direct-importer-for-mac-app-store-build

harr1424 committed:
refactor per ai review

harr1424 committed:
use i18n for NSOpenPanel dialog

coroiu reviewed (approved):
No changes to platform code
